staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go

```diff
@@ -113,71 +114,132 @@ func WithPriorityAndFairness(
 				waitingMark.recordReadOnly(int(atomic.AddInt32(&atomicReadOnlyWaiting, delta)))
 			}
 		}
-		var resultCh chan interface{}
-		var forgetWatch utilflowcontrol.ForgetWatchFunc
-		if isWatchRequest {
-			resultCh = make(chan interface{})
+		queueNote := func(inQueue bool) {
+			if inQueue {
+				noteWaitingDelta(1)
+			} else {
+				noteWaitingDelta(-1)
+			}
 		}
-		execute := func() {
-			noteExecutingDelta(1)
-			defer noteExecutingDelta(-1)
-			served = true
 
-			innerCtx := ctx
-			innerReq := r
+		// find the estimated "width" of the request
+		// TODO: Maybe just make it costEstimator and let it return additionalLatency too for the watch?
+		// TODO: Estimate cost should also take fcIfc.GetWatchCount(requestInfo) as a parameter.
+		width := widthEstimator.EstimateWidth(r)
+		digest := utilflowcontrol.RequestDigest{RequestInfo: requestInfo, User: user, Width: width}
 
+		if isWatchRequest {
+			shouldStartWatchCh := make(chan struct{})
 			var watchInitializationSignal utilflowcontrol.InitializationSignal
```

Comment:
I wanted to avoid creating those to avoid unnecessary memory allocations if those aren't needed.
[Same below]

If you think it's overoptimization, I can change that, but for now leaving it as is.

Reply:
the signal is relatively cheap… I think the clarity is worth it

Reply:
makes sense; fixed